File: tauri-app/src/features/connection/utils/connectionError.ts
import type { Composer } from 'vue-i18n'

/** 帮助文档链接（按错误类型针对性跳转） */
const QUICK_START_URL = 'https://micyou.top/docs/quick-start'
const FAQ_URL = 'https://micyou.top/docs/faq'

export type ConnectionErrorType =
  | 'NetworkTimeout'
  | 'NetworkUnreachable'
  | 'PortInUse'
  | 'ConnectionRefused'
  | 'PermissionDenied'
  | 'FirewallBlocked'
  | 'AdminPrivilegeRequired'
  | 'DeviceNotFound'
  | 'UsbConnectionFailed'
  | 'AdbCommandFailed'
  | 'HandshakeFailed'
  | 'ProtocolError'
  | 'VersionMismatch'
  | 'UdpPortBlocked'
  | 'AudioDeviceError'
  | 'AudioFormatError'
  | 'UnknownError'

export interface ConnectionErrorDetails {
  type: ConnectionErrorType
  originalMessage: string
  title: string
  message: string
  suggestions: string[]
  showRetry: boolean
  showHelp: boolean
  helpUrl: string | null
}

export function analyzeError(errorMessage: string): ConnectionErrorType {
  const msg = errorMessage.toLowerCase()

  if (msg.includes('timeout')) return 'NetworkTimeout'
  if (msg.includes('bind') || msg.includes('port is already in use') || msg.includes('address already in use')) return 'PortInUse'
  if (msg.includes('connection refused') || msg.includes('refused')) return 'ConnectionRefused'
  if (msg.includes('unreachable') || msg.includes('no route to host') || msg.includes('network is unreachable')) return 'NetworkUnreachable'
  if (msg.includes('firewall') || msg.includes('blocked')) return 'FirewallBlocked'
  if (msg.includes('permission') || msg.includes('access denied') || msg.includes('privilege')) return 'PermissionDenied'
  if (msg.includes('adb')) return 'AdbCommandFailed'
  if (msg.includes('usb')) return 'UsbConnectionFailed'
  if (msg.includes('handshake') || msg.includes('握手')) return 'HandshakeFailed'
  if (msg.includes('audio')) return 'AudioDeviceError'
  if (msg.includes('udp')) return 'UdpPortBlocked'

  return 'UnknownError'
}

function extractAdbCommand(message: string): string | null {
  for (const delimiter of ['：', ':']) {
    const idx = message.indexOf(delimiter)
    if (idx !== -1) {
      const after = message.substring(idx + 1).trim()
      if (after) return after
    }
  }
  return null
}

export function generateErrorDetails(
  type: ConnectionErrorType,
  originalMessage: string,
  mode: string,
  port: number | undefined,
  ip: string | undefined,
  t: Composer['t']
): ConnectionErrorDetails {

  const base = {
    type,
    originalMessage,
    showRetry: true,
    showHelp: false,
    helpUrl: null as string | null,
  }

  switch (type) {
    case 'NetworkTimeout':
      return {
        ...base,
        title: t('error.networkTimeout.title'),
        message: t('error.networkTimeout.message'),
        suggestions: [
          t('error.suggestion.checkNetwork'),
          t('error.suggestion.checkTargetRunning'),
          t('error.suggestion.tryDifferentPort'),
        ],
      }

    case 'PortInUse':
      return {
        ...base,
        title: t('error.portInUse.title'),
        message: t('error.portInUse.message', { port: String(port ?? 8554) }),
        suggestions: [
          t('error.suggestion.changePort'),
          t('error.suggestion.checkOtherApps'),
        ],
      }

    case 'ConnectionRefused':
      return {
        ...base,
        title: t('error.connectionRefused.title'),
        message: mode === 'wifi'
          ? t('error.connectionRefused.wifiMessage', { ip: ip ?? '' })
          : t('error.connectionRefused.message'),
        suggestions: [
          t('error.suggestion.checkServerRunning'),
          t('error.suggestion.checkServerConfig'),
        ],
      }

    case 'NetworkUnreachable':
      return {
        ...base,
        title: t('error.networkUnreachable.title'),
        message: t('error.networkUnreachable.message', { ip: ip ?? '' }),
        suggestions: [
          t('error.suggestion.checkNetworkConnection'),
          t('error.suggestion.verifyIpAddress'),
          t('error.suggestion.checkWifiConnected'),
        ],
      }

    case 'FirewallBlocked':
      return {
        ...base,
        title: t('error.firewallBlocked.title'),
        message: t('error.firewallBlocked.message', { port: String(port ?? 8554) }),
        suggestions: [
          t('error.suggestion.addFirewallRule'),
          t('error.suggestion.runAsAdmin'),
        ],
        showHelp: true,
        helpUrl: FAQ_URL,
      }

    case 'PermissionDenied':
      return {
        ...base,
        title: t('error.permissionDenied.title'),
        message: t('error.permissionDenied.message'),
        suggestions: [
          t('error.suggestion.runAsAdmin'),
          t('error.suggestion.checkAntivirus'),
        ],
      }

    case 'DeviceNotFound':
      return {
        ...base,
        title: t('error.deviceNotFound.title'),
        message: t('error.deviceNotFound.message'),
        suggestions: [t('error.suggestion.checkNetworkConnection')],
      }

    case 'UsbConnectionFailed':
      return {
        ...base,
        title: t('error.usbConnectionFailed.title'),
        message: t('error.usbConnectionFailed.message'),
        suggestions: [
          t('error.suggestion.checkUsbCable'),
          t('error.suggestion.enableUsbDebugging'),
          ...(extractAdbCommand(originalMessage)
            ? [t('error.suggestion.runAdbCommand', { cmd: extractAdbCommand(originalMessage)! })]
            : []),
        ],
        showHelp: true,
        helpUrl: FAQ_URL,
      }

    case 'AdbCommandFailed':
      return {
        ...base,
        title: t('error.adbCommandFailed.title'),
        message: t('error.adbCommandFailed.message'),
        suggestions: [
          t('error.suggestion.checkAdbInstalled'),
          ...(extractAdbCommand(originalMessage)
            ? [t('error.suggestion.runAdbManually', { cmd: extractAdbCommand(originalMessage)! })]
            : []),
        ],
        showHelp: true,
        helpUrl: QUICK_START_URL,
      }

    case 'HandshakeFailed':
      return {
        ...base,
        title: t('error.handshakeFailed.title'),
        message: t('error.handshakeFailed.message'),
        suggestions: [
          t('error.suggestion.versionMatch'),
          t('error.suggestion.restartApp'),
        ],
      }

    case 'ProtocolError':
      return {
        ...base,
        title: t('error.protocolError.title'),
        message: t('error.protocolError.message'),
        suggestions: [
          t('error.suggestion.restartApp'),
          t('error.suggestion.checkVersion'),
        ],
      }

    case 'AudioDeviceError':
      return {
        ...base,
        title: t('error.audioDevice.title'),
        message: t('error.audioDevice.message'),
        suggestions: [
          t('error.suggestion.checkAudioDevice'),
          t('error.suggestion.restartApp'),
        ],
      }

    case 'AudioFormatError':
      return {
        ...base,
        title: t('error.audioFormat.title'),
        message: t('error.audioFormat.message'),
        suggestions: [
          t('error.suggestion.changeAudioConfig'),
          t('error.suggestion.useDefaultConfig'),
        ],
      }

    case 'VersionMismatch':
      return {
        ...base,
        title: t('error.versionMismatch.title'),
        message: t('error.versionMismatch.message'),
        suggestions: [
          t('error.suggestion.updateApp'),
          t('error.suggestion.checkVersion'),
        ],
        showHelp: true,
        helpUrl: QUICK_START_URL,
      }

    case 'AdminPrivilegeRequired':
      return {
        ...base,
        title: t('error.adminPrivilege.title'),
        message: t('error.adminPrivilege.message'),
        suggestions: [t('error.suggestion.runAsAdmin')],
      }

    case 'UdpPortBlocked':
      return {
        ...base,
        title: t('error.udpPortBlocked.title'),
        message: t('error.udpPortBlocked.message', { port: port ? port + 1 : 6001 }),
        suggestions: [
          t('error.suggestion.addFirewallRule'),
          t('error.suggestion.runAsAdmin'),
        ],
        showHelp: true,
        helpUrl: FAQ_URL,
      }

    case 'UnknownError':
    default:
      return {
        ...base,
        title: t('error.unknown.title'),
        message: t('error.unknown.message', { error: originalMessage }),
        suggestions: [
          t('error.suggestion.restartApp'),
          t('error.suggestion.checkLogs'),
        ],
        showHelp: true,
        helpUrl: FAQ_URL,
      }
  }
}

File: composeApp/src/main/kotlin/com/lanrhyme/micyou/network/ConnectionError.kt
package com.lanrhyme.micyou.network

import com.lanrhyme.micyou.R
import com.lanrhyme.micyou.util.Constants
import com.lanrhyme.micyou.util.getString
import com.lanrhyme.micyou.viewmodel.ConnectionMode

/**
 * 连接错误类型枚举
 * 用于分类不同的连接失败原因，提供更精确的错误信息和恢复建议
 */
enum class ConnectionErrorType {
    // 网络连接相关错误
    NetworkTimeout,          // 网络连接超时
    NetworkUnreachable,      // 网络不可达（IP 错误或网络断开）
    PortInUse,               // 端口已被占用
    ConnectionRefused,       // 连接被拒绝（服务未启动）

    // 权限相关错误
    PermissionDenied,        // 权限不足
    VpnBlocked,              // VPN 防火墙拦截了局域网直连流量

    // 设备相关错误
    DeviceNotFound,          // 设备未找到
    UsbConnectionFailed,     // USB 连接失败
    AdbCommandFailed,        // ADB 命令执行失败

    // 协议相关错误
    HandshakeFailed,         // 握手失败（协议不匹配）
    ProtocolError,           // 协议错误
    VersionMismatch,         // 版本不匹配

    // UDP 相关错误
    UdpPortBlocked,          // UDP 端口被阻止

    // 音频相关错误
    AudioDeviceError,        // 音频设备错误
    AudioFormatError,        // 音频格式不支持

    // 通用错误
    UnknownError             // 未知的错误类型
}

/**
 * 连接错误详情
 * 包含错误类型、原始错误消息、恢复建议等详细信息
 */
data class ConnectionErrorDetails(
    val type: ConnectionErrorType,
    val originalMessage: String,
    val localizedTitle: String,
    val localizedMessage: String,
    val recoverySuggestions: List<String> = emptyList(),
    val showRetryButton: Boolean = true,
    val showHelpButton: Boolean = false,
    val helpUrl: String? = null
)

/**
 * 连接错误助手类
 * 用于分析和生成详细的错误信息
 */
object ConnectionErrorHelper {
    
    /**
     * 根据异常分析错误类型
     */
    fun analyzeError(exception: Exception, mode: ConnectionMode): ConnectionErrorType {
        val message = exception.message ?: ""
        
        return when {
            // VPN kill-switch style firewall rejecting LAN-bound traffic (EPERM)
            message.contains("not permitted", ignoreCase = true) ||
            message.contains("eperm", ignoreCase = true) ->
                ConnectionErrorType.VpnBlocked

            // 网络超时
            message.contains("timeout", ignoreCase = true) ||
            message.contains("Timeout", ignoreCase = true) ->
                ConnectionErrorType.NetworkTimeout
            
            // 端口占用
            message.contains("Bind", ignoreCase = true) ||
            message.contains("port is already in use", ignoreCase = true) ||
            message.contains("Address already in use", ignoreCase = true) ->
                ConnectionErrorType.PortInUse
            
            // 连接被拒绝
            message.contains("Connection refused", ignoreCase = true) ||
            message.contains("refused", ignoreCase = true) ->
                ConnectionErrorType.ConnectionRefused
            
            // 网络不可达
            message.contains("unreachable", ignoreCase = true) ||
            message.contains("No route to host", ignoreCase = true) ||
            message.contains("Network is unreachable", ignoreCase = true) ->
                ConnectionErrorType.NetworkUnreachable
            
            // 权限不足
            message.contains("permission", ignoreCase = true) ||
            message.contains("access denied", ignoreCase = true) ||
            message.contains("privilege", ignoreCase = true) ->
                ConnectionErrorType.PermissionDenied
            
            // ADB 相关
            message.contains("adb", ignoreCase = true) ->
                ConnectionErrorType.AdbCommandFailed
            
            // USB 相关
            message.contains("usb", ignoreCase = true) ||
            message.contains("USB", ignoreCase = true) ->
                ConnectionErrorType.UsbConnectionFailed
            
            // 握手失败
            message.contains("handshake", ignoreCase = true) ||
            message.contains("握手", ignoreCase = true) ->
                ConnectionErrorType.HandshakeFailed
            
            // 音频相关
            message.contains("audio", ignoreCase = true) ||
            message.contains("Audio", ignoreCase = true) ->
                ConnectionErrorType.AudioDeviceError
            
            // UDP 相关
            message.contains("udp", ignoreCase = true) ||
            message.contains("UDP", ignoreCase = true) ->
                ConnectionErrorType.UdpPortBlocked
            
            // 其他
            else -> ConnectionErrorType.UnknownError
        }
    }
    
    private fun extractAdbCommand(message: String): String? {
        val delimiters = listOf("：", ":")
        for (delimiter in delimiters) {
            val afterDelimiter = message.substringAfter(delimiter).trim()
            if (afterDelimiter.isNotBlank() && afterDelimiter != message) {
                return afterDelimiter
            }
        }
        return null
    }
    
    /**
     * 生成详细的错误信息（需要配合 Localization）
     */
    suspend fun generateErrorDetails(
        type: ConnectionErrorType,
        originalMessage: String,
        mode: ConnectionMode,
        port: Int? = null,
        ip: String? = null
    ): ConnectionErrorDetails {
        return when (type) {
            ConnectionErrorType.NetworkTimeout -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorNetworkTimeoutTitle),
                localizedMessage = getString(R.string.errorNetworkTimeoutMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionCheckNetwork),
                    getString(R.string.errorSuggestionCheckTargetRunning),
                    getString(R.string.errorSuggestionTryDifferentPort)
                )
            )
            
            ConnectionErrorType.PortInUse -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorPortInUseTitle),
                localizedMessage = String.format(getString(R.string.errorPortInUseMessage), port?.toString() ?: Constants.DEFAULT_TCP_PORT.toString()),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionChangePort),
                    getString(R.string.errorSuggestionCheckOtherApps)
                )
            )
            
            ConnectionErrorType.ConnectionRefused -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorConnectionRefusedTitle),
                localizedMessage = if (mode == ConnectionMode.Wifi) 
                    String.format(getString(R.string.errorConnectionRefusedWifiMessage), ip ?: "")
                else getString(R.string.errorConnectionRefusedMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionCheckServerRunning),
                    getString(R.string.errorSuggestionCheckServerConfig)
                )
            )
            
            ConnectionErrorType.NetworkUnreachable -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorNetworkUnreachableTitle),
                localizedMessage = String.format(getString(R.string.errorNetworkUnreachableMessage), ip ?: ""),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionCheckNetworkConnection),
                    getString(R.string.errorSuggestionVerifyIpAddress),
                    getString(R.string.errorSuggestionCheckWifiConnected)
                )
            )
            
            ConnectionErrorType.PermissionDenied -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorPermissionDeniedTitle),
                localizedMessage = getString(R.string.errorPermissionDeniedMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionCheckSettings)
                )
            )

            ConnectionErrorType.VpnBlocked -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorVpnBlockedTitle),
                localizedMessage = getString(R.string.errorVpnBlockedMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionVpnAllowLan),
                    getString(R.string.errorSuggestionVpnDisableAlwaysOn),
                    getString(R.string.errorSuggestionVpnTemporarilyOff)
                )
            )
            
            ConnectionErrorType.DeviceNotFound -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorDeviceNotFoundTitle),
                localizedMessage = getString(R.string.errorDeviceNotFoundMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionCheckNetworkConnection)
                )
            )

            ConnectionErrorType.UsbConnectionFailed -> {
                val command = extractAdbCommand(originalMessage)
                ConnectionErrorDetails(
                    type = type,
                    originalMessage = originalMessage,
                    localizedTitle = getString(R.string.errorUsbConnectionFailedTitle),
                    localizedMessage = getString(R.string.errorUsbConnectionFailedMessage),
                    recoverySuggestions = buildList {
                        add(getString(R.string.errorSuggestionCheckUsbCable))
                        add(getString(R.string.errorSuggestionEnableUsbDebugging))
                        if (command != null) {
                            add(String.format(getString(R.string.errorSuggestionRunAdbCommand), command))
                        }
                    },
                    showHelpButton = true,
                    helpUrl = "https://github.com/LanRhyme/MicYou/blob/master/docs/FAQ.md#usb"
                )
            }
            
            ConnectionErrorType.AdbCommandFailed -> {
                val command = extractAdbCommand(originalMessage)
                ConnectionErrorDetails(
                    type = type,
                    originalMessage = originalMessage,
                    localizedTitle = getString(R.string.errorAdbCommandFailedTitle),
                    localizedMessage = getString(R.string.errorAdbCommandFailedMessage),
                    recoverySuggestions = buildList {
                        add(getString(R.string.errorSuggestionCheckAdbInstalled))
                        if (command != null) {
                            add(String.format(getString(R.string.errorSuggestionRunAdbManually), command))
                        }
                    },
                    showHelpButton = true,
                    helpUrl = "https://github.com/LanRhyme/MicYou/blob/master/docs/FAQ.md#usb"
                )
            }
            
            ConnectionErrorType.HandshakeFailed -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorHandshakeFailedTitle),
                localizedMessage = getString(R.string.errorHandshakeFailedMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionVersionMatch),
                    getString(R.string.errorSuggestionRestartApp)
                )
            )
            
            ConnectionErrorType.ProtocolError -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorProtocolErrorTitle),
                localizedMessage = getString(R.string.errorProtocolErrorMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionRestartApp),
                    getString(R.string.errorSuggestionCheckVersion)
                )
            )
            
            ConnectionErrorType.AudioDeviceError -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorAudioDeviceTitle),
                localizedMessage = getString(R.string.errorAudioDeviceMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionCheckAudioDevice),
                    getString(R.string.errorSuggestionRestartApp)
                )
            )
            
            ConnectionErrorType.AudioFormatError -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorAudioFormatTitle),
                localizedMessage = getString(R.string.errorAudioFormatMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionChangeAudioConfig),
                    getString(R.string.errorSuggestionUseDefaultConfig)
                )
            )
            
            ConnectionErrorType.VersionMismatch -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorVersionMismatchTitle),
                localizedMessage = getString(R.string.errorVersionMismatchMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionUpdateApp),
                    getString(R.string.errorSuggestionCheckVersion)
                ),
                showHelpButton = true,
                helpUrl = "https://github.com/LanRhyme/MicYou/releases"
            )
            
            ConnectionErrorType.UnknownError -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorUnknownTitle),
                localizedMessage = String.format(getString(R.string.errorUnknownMessage), originalMessage),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionRestartApp),
                    getString(R.string.errorSuggestionCheckLogs)
                ),
                showHelpButton = true,
                helpUrl = "https://github.com/LanRhyme/MicYou/issues"
            )
            
            ConnectionErrorType.UdpPortBlocked -> ConnectionErrorDetails(
                type = type,
                originalMessage = originalMessage,
                localizedTitle = getString(R.string.errorUdpPortBlockedTitle),
                localizedMessage = getString(R.string.errorUdpPortBlockedMessage, port?.let { calculateUdpPort(it) } ?: Constants.DEFAULT_UDP_PORT),
                recoverySuggestions = listOf(
                    getString(R.string.errorSuggestionCheckNetwork),
                    getString(R.string.errorSuggestionRestartApp)
                ),
                showHelpButton = true,
                helpUrl = "https://github.com/LanRhyme/MicYou/blob/master/docs/FAQ.md"
            )
        }
    }
}